npm/pkg/dataplane/dpshim/dpshim.go

```diff
@@ -1,77 +1,590 @@
+		if !dp.policyExists(policyKey) {
+			return nil, npmerrors.Errorf(npmerrors.AddPolicy, false, fmt.Sprintf("policy %s not found", policyKey))
+		}
+
+		policy := dp.policyCache[policyKey]
+		toApplyPolicies[idx] = policy
+		idx++
+	}
+
+	payload, err := controlplane.EncodeNPMNetworkPolicies(toApplyPolicies)
+	if err != nil {
+		klog.Errorf("processPoliciesApply: failed to encode policies %v", err)
+		return nil, npmerrors.ErrorWrapper(npmerrors.AddPolicy, false, "processPoliciesApply: failed to encode sets", err)
+	}
+
+	return getGoalStateFromBuffer(payload), nil
+}
+
+func (dp *DPShim) processPoliciesRemove() (*protos.GoalState, error) {
+	if len(dp.dirtyCache.toDeletePolicies) == 0 {
+		return nil, nil
+	}
+
+	toDeletePolicies := make([]string, len(dp.dirtyCache.toDeletePolicies))
+	idx := 0
+
+	for policyKey := range dp.dirtyCache.toDeletePolicies {
+		toDeletePolicies[idx] = policyKey
+		idx++
+	}
+
+	payload, err := controlplane.EncodeStrings(toDeletePolicies)
+	if err != nil {
+		klog.Errorf("processPoliciesRemove: failed to encode policies %v", err)
+		return nil, npmerrors.ErrorWrapper(npmerrors.RemovePolicy, false, "processPoliciesRemove: failed to encode sets", err)
+	}
+
+	return getGoalStateFromBuffer(payload), nil
+}
+
+func (dp *DPShim) hydrateSetCache() (*protos.GoalState, error) {
+	if len(dp.setCache) == 0 {
+		return nil, nil
+	}
+	toApplySets := make([]*controlplane.ControllerIPSets, len(dp.setCache))
+	idx := 0
+
+	for _, set := range dp.setCache {
+		toApplySets[idx] = set
+		idx++
+	}
+
+	payload, err := controlplane.EncodeControllerIPSets(toApplySets)
+	if err != nil {
+		klog.Errorf("processIPSetsApply: failed to encode sets %v", err)
+		return nil, npmerrors.ErrorWrapper(npmerrors.AppendIPSet, false, "processIPSetsApply: failed to encode sets", err)
+	}
+
+	return getGoalStateFromBuffer(payload), nil
+}
+
+func (dp *DPShim) hydratePolicyCache() (*protos.GoalState, error) {
+	if len(dp.policyCache) == 0 {
+		return nil, nil
+	}
+
+	toApplyPolicies := make([]*policies.NPMNetworkPolicy, len(dp.policyCache))
+	idx := 0
+
+	for _, policy := range dp.policyCache {
+		toApplyPolicies[idx] = policy
+		idx++
+	}
+
+	payload, err := controlplane.EncodeNPMNetworkPolicies(toApplyPolicies)
+	if err != nil {
+		klog.Errorf("processPoliciesApply: failed to encode policies %v", err)
+		return nil, npmerrors.ErrorWrapper(npmerrors.AddPolicy, false, "processPoliciesApply: failed to encode sets", err)
+	}
+
+	return getGoalStateFromBuffer(payload), nil
+}
+
+func (dp *DPShim) deleteUnusedSets(stopChannel <-chan struct{}) {
+	go func() {
+		ticker := time.NewTicker(time.Hour * time.Duration(cleanEmptySetsInHrs))
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-stopChannel:
+				return
+			case <-ticker.C:
+				klog.Info("deleteUnusedSets: cleaning up unused sets")
+				dp.checkSetReferences()
+				err := dp.ApplyDataPlane()
+				if err != nil {
+					klog.Errorf("deleteUnusedSets: failed to apply dataplane %v", err)
```

Comment:
I've been curious about error propagation in go routines. Should we create an error channel to feedback errors and handle them in the caller code?

Reply:
This is across our codebase I mean